plugins/cros_ec-usb/fu-cros_ec-usb-device.c

```diff
@@ -0,0 +1,327 @@
+
+	if (self->protocol_version < 5 || self->protocol_version > 6) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "Unsupported protocol version %d: \n",
+			     self->protocol_version);
+		return FALSE;
+	}
+	self->header_type = GUINT16_FROM_BE (start_resp.rpdu.header_type);
+
+	error_code = GUINT32_FROM_BE (start_resp.rpdu.return_value);
+	if (error_code) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "Target reporting error %u: \n", error_code);
+		return FALSE;
+	}
+
+	memcpy (self->targ.common.version, start_resp.rpdu.common.version,
+		sizeof(start_resp.rpdu.common.version));
+	self->targ.common.maximum_pdu_size =
+		GUINT32_FROM_BE (start_resp.rpdu.common.maximum_pdu_size);
+	self->targ.common.flash_protection =
+		GUINT32_FROM_BE (start_resp.rpdu.common.flash_protection);
+	self->targ.common.min_rollback = GINT32_FROM_BE (start_resp.rpdu.common.min_rollback);
+	self->targ.common.key_version = GUINT32_FROM_BE (start_resp.rpdu.common.key_version);
+
+	if (!fu_cros_ec_usb_parse_version (self->targ.common.version,
+					   &self->version, error)) {
+		g_prefix_error (error,
+				"Failed parsing device's version: %32s\n",
+				self->targ.common.version);
+		return FALSE;
+	}
+
+	fu_device_set_version (FU_DEVICE(device), self->version.triplet);
+	fu_device_add_instance_id (FU_DEVICE(device), self->version.boardname);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_close (FuUsbDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	if (!g_usb_device_release_interface (usb_device, self->iface_idx,
+					     G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					     error)) {
+		g_prefix_error (error, "failed to release interface: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_cros_ec_usb_device_init (FuCrOSECUsbDevice *device)
+{
+	fu_device_set_version_format (FU_DEVICE (device), FWUPD_VERSION_FORMAT_TRIPLET);
+}
+
+
+static void
+fu_cros_ec_usb_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	g_autofree gchar *flash_protection = NULL;
+	g_autofree gchar *min_rollback = NULL;
+
+	fu_common_string_append_kv (str, idt, "GitHash", self->version.sha1);
+	fu_common_string_append_kb (str, idt, "Dirty",
+				    self->version.dirty);
+	fu_common_string_append_ku (str, idt, "ProtocolVersion",
+				    self->protocol_version);
+	fu_common_string_append_ku (str, idt, "HeaderType",
+				    self->header_type);
+	fu_common_string_append_ku (str, idt, "MaxPDUSize",
+				    self->targ.common.maximum_pdu_size);
+	flash_protection = g_strdup_printf ("%04x",
+					    self->targ.common.flash_protection);
+	fu_common_string_append_kv (str, idt, "FlashProtectionStatus",
+				    flash_protection);
```

Comment:
I think you can use fu_common_string_append_kx for this.

Reply:
Done.